### PR TITLE
Added bus interface

### PR DESCRIPTION
**Pull request by ashcrow**
Handlers now are passed a bus input. If a bus is configured, the input will be an instance of ```commissaire_http.bus.Bus``` which provides a connection and methods to the bus to send and receive messages.

Docs are at https://github.com/projectatomic/commissaire/pull/2

### CODE REVIEW

File: src/commissaire_http/bus/__init__.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            'qkwargs': self.qkwargs,
+        }
+
+    def connect(self):
+        """
+        'Connects' to the bus.
+
+        :returns: The same instance.
+        :rtype: commissaire_http.bus.Bus
+        """
```
Comment (mbarnes) [inline src/commissaire_http/bus/__init__.py]: Probably over-thinking at this early point, but the questions occurred to me so I'll ask them:
 * Would it be useful to make this method idempotent (based on `self.connection == None`)?
 * Should we catch exceptions and reset internal members back to their initial state so we don't leave the bus instance half-connected?

File: src/commissaire_http/bus/__init__.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+    @classmethod
+    def create_id(cls):  # pragma: no cover
+        """
+        Creates a new unique identifier.
+
+        :returns: A unique identification string.
+        :rtype: str
+        """
+        return str(uuid.uuid4())
```
Comment (mbarnes) [inline src/commissaire_http/bus/__init__.py]: See my comment below about a mixin class in `commissaire`.  This method could also be moved there.

File: src/commissaire_http/bus/__init__.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        send_queue = self.connection.SimpleQueue(queue_name, **kwargs)
+        jsonrpc_msg = {
+            'jsonrpc': "2.0",
+            'id': id,
+            'result': payload,
+        }
+        self.logger.debug('jsonrpc msg: {}'.format(jsonrpc_msg))
+        send_queue.put(jsonrpc_msg)
+        self.logger.debug('Sent response for message id "{}"'.format(id))
+        send_queue.close()
```
Comment (mbarnes) [inline src/commissaire_http/bus/__init__.py]: When would `respond()` be used on the HTTP side?

Also, `CommissaireService` has an identical method.  Maybe have a mixin class in `commissaire` for use by both `CommissaireService` and `Bus`?

File: src/commissaire_http/bus/__init__.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            self.logger.warn(
+                'Error returned from the message id "{}"'.format(
+                    id, result.payload))
+
+        self.logger.debug(
+            'Result retrieved from response queue "{}": payload="{}"'.format(
+                response_queue_name, result))
+        self.logger.debug('Closing queue {}'.format(response_queue_name))
+        response_queue.close()
+        return result.payload
```
Comment (mbarnes) [inline src/commissaire_http/bus/__init__.py]: `CommissaireService` has an identical method.  Maybe have a mixin class in `commissaire` for use by both `CommissaireService` and `Bus`?

File: src/commissaire_http/__init__.py [hunk 5]
```diff
@@ -30,25 +32,39 @@ class ThreadedWSGIServer(ThreadingMixIn, WSGIServer):
     pass
 
 
+class CommisaireRequestHandler(WSGIRequestHandler):
```
Comment (ashcrow) [inline src/commissaire_http/__init__.py]: :scream:

File: src/commissaire_http/bus/__init__.py (hunk 3 above)
Comment (ashcrow) [inline src/commissaire_http/bus/__init__.py]: That's the plan! I want to move some of the items used in both places in to the ``comissaire`` package.

File: src/commissaire_http/bus/__init__.py (hunk 2 above)
Comment (ashcrow) [inline src/commissaire_http/bus/__init__.py]: :+1: but I'd like to do it later.

File: src/commissaire_http/bus/__init__.py (hunk 1 above)
Comment (ashcrow) [inline src/commissaire_http/bus/__init__.py]: > Would it be useful to make this method idempotent (based on self.connection == None)?

Can't hurt. It's a small change so I'll do it now.

> Should we catch exceptions and reset internal members back to their initial state so we don't leave the bus instance half-connected?

Good idea. Let's tackle that in follow up work.

**Comment by ashcrow** [comment]
Hrm. I meant to have this against ``master``. Ignore the conflict. If things look good I'll switch the branch to point to master before merge.

**Comment by ashcrow** [comment]
:arrow_up:

**Comment by mbarnes** [comment]
Looks good!  :+1:  Gotta run, but I'll merge this as soon as I can.

**Comment by mbarnes** [comment]
Squashed and merged manually.

**Pull request closed without merging**